### PR TITLE
Enthalpy injection dirac kernel

### PR DESCRIPTION
**Pull request by andrsd**
Adding a dirac kernel for enthalpy injection. Users need to provide
temperature computed by a postprocessor, pressure and fluid properties.
This object should be used in conjunction with PorousFlowPointSourceFromPostprocessor

Refs #15742

### CODE REVIEW

**Comment by WilkAndy** [comment]
Yep, i can review, @andrsd .   Please give me a couple of days and then remind me if i haven't done it!

**Comment by andrsd** [comment]
just a friendly reminder @WilkAndy ...

**Comment by WilkAndy** [comment]
Aaah, i'm really under the pump at the moment.  Keep reminding me, please!

a

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/dirackernels/hfrompps.i:245]: You should also have a `PorousFlowHeatEnergy` Postprocessor so that your gold file demonstrates the correct increasing value of fluid heat energy

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/dirackernels/hfrompps.i:237]: Probably don't want "mumps" to be used as the linear solver, because some systems might not have it.

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/dirackernels/hfrompps.i:279]: `perf_graph` probably not needed

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:23]: I've written heaps of comments, but this is the main one.... Is "pressure" supposed to be (1) the pressure of the fluid in the porous material, or (2) the pressure of the injected fluid. 

(1) You should get pressure from the PorousFlow material system, and you should provide derivatives of the residual in a computeQpJacobian.

(2) You should state in the documentation (param and markdown) that this is usually an AuxVariable representing the pressure of an external agent, such as a borehole, but that if it is a PorousFlow variable then no Jacobian entries are computed so this DiracKernel may result in slightly slower numerical convergence.

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:21]: "only positive values are allowed".  You don't check for positivity.  Do you want to?  Or do you want to change to "please ensure this is positive so that this object acts like a source"?

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:22]: "for fluid properties" -> "used to calculate the fluid properties of the injected fluid"

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:20]: "mass flux" -> "mass flux of injected fluid"

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:25]: "temperature of fluid" -> "temperature of injected fluid"

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [inline modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:28]: Point source that adds heat energy corresponding to injection of a fluid with specified mass flux rate (specified by a postprocessor) at given temperature (specified by a postprocessor)

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (WilkAndy) [inline modules/porous_flow/test/tests/dirackernels/hfrompps.i:207]: If you eliminate the advection Kernels you won't need these Materials

File: modules/porous_flow/doc/content/modules/porous_flow/sinks.md
Comment (WilkAndy) [inline modules/porous_flow/doc/content/modules/porous_flow/sinks.md:36]: I think you should replace "users have to add" by "users can add".  Then after the sentence "Users can also fix the temperature of an injected fluid using a Dirichlet BC, but this adds/subtracts heat energy to entire nodal volumes of porous material and fluid, so may lead to unacceptable errors from the additional heat energies added/removed"

**Review by WilkAndy — CHANGES_REQUESTED** [review]

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (andrsd) [reply modules/porous_flow/test/tests/dirackernels/hfrompps.i:83]: Yeah, I was primarily focusing on testing the different temperature, but I can change mass flux as well. Probably a good idea :+1:

File: modules/porous_flow/doc/content/source/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.md
Comment (andrsd) [reply modules/porous_flow/doc/content/source/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.md:5]: Mentioning the requirement to use the identical fluid property object makes sense to me.  It seems misleading to mention the materials, since they are not really needed in this Dirac kernel setup... Unless you have some other reason to mention them.

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (andrsd) [reply modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:21]: I don't want to check positivity at the moment. It happens to us that for very short time the mass flux is actually negative. Eventually, I would like to add logic that would look at the direction and did the right thing.

So I will change the doco.

**Comment by andrsd** [comment]
I tried the suggestion to remove the BCs and advection kernels, but then I am getting zero eigenvalues on the Jacobian matrix, and hence no covergence.

File: modules/porous_flow/test/tests/dirackernels/hfrompps.i
Comment (andrsd) [reply modules/porous_flow/test/tests/dirackernels/hfrompps.i:237]: Gone

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (andrsd) [reply modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:26]: Are you saying that the pressure value I am using is not the right one? Because I do use the pressure we solve for, but seems like it should be the pressure of the injected material. If that is so, then does that need to be a user input instead?

**Comment by andrsd** [comment]
Ok. I managed to remove the BCs. They were needed, because I was getting non-physical states in the IAPWS97 fluid properties. I switched to simple fluid and they are not needed.

But, I was not able to remove the advection kernels. Still singular eigenvalues in the Jacobian.

**Comment by andrsd** [comment]
@WilkAndy could you look at this when you get some time, please?

**Comment by WilkAndy** [comment]
Oh bum, i forgot about it!    Yes, i hope to look at it today,

a

File: modules/porous_flow/include/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.h
Comment (WilkAndy) [inline modules/porous_flow/include/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.h:18]: Point source that adds heat energy corresponding to adding a fluid with flux rate specified by a postprocessor at given temperature (specified by a postprocessor).

File: modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C
Comment (WilkAndy) [reply modules/porous_flow/src/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.C:26]: Upon a bit more reflection, i think you are potentially doing (1) and (2).   So i think the documentation (in param and markdown) should state this.  Something like:

"Pressure (along with T_in) is used to calculate the injected fluid enthalpy.  Most frequently, it is the PorousFlow pressure variable (the porepressure in the porous medium).  This models the situation where fluid is injected at a specified rate and temperature (using this DiracKernel and a PorousFlowPointSourceFromPostprocessor) which potentially leads to changes in the porepressure.  Alternately, it may be the pressure of the injected fluid, as fixed by an external agent (such as a pump) which is stored in an AuxVariable.

**Review by WilkAndy — CHANGES_REQUESTED** [review]
It be great to have a Jacobian test for this, but i think you might be wanting to finalise this chunk of work.  In which case, just change the doco similarly to what i suggest and merge the sucker.

**Comment by andrsd** [comment]
> It be great to have a Jacobian test for this

Let's see if I can build one. Should be a big deal - we do that a lot in other apps I oversee.

**Comment by andrsd** [comment]
Thanks for pointing out the Jacobian test - it uncovered a bug in my Jacobian statement.

**Comment by WilkAndy** [comment]
Thanks for doing this work, @andrsd .   I just wanted to confirm that i'm perfectly happy with it, and it can be merged as far as i'm concerned.  I don't know how to fix your documentation error, sorry.

**Comment by andrsd** [comment]
> I don't know how to fix your documentation error, sorry.

I had `PorousFlowPointSourceFromPostprocessor` type in the param spec instead of `PorousFlowPointEnthalpySourceFromPostprocessor`. Dumb copy-pasta error.

**Comment by WilkAndy** [comment]
This PR can be merged, i believe, @aeslaughter

**Comment by andrsd** [comment]
It needs an approving review from somebody. Do you not have the powers @WilkAndy ?

File: modules/porous_flow/include/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.h
Comment (loganharbour) [inline modules/porous_flow/include/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.h:27]: Who calls this?

Comment (loganharbour) [reply modules/porous_flow/include/dirackernels/PorousFlowPointEnthalpySourceFromPostprocessor.h:27]: Disregard. I see that it's virtual

**Review by loganharbour — APPROVED** [review]
Looks good. Please hold off if you can for when next is passing again

**Comment by andrsd** [comment]
Thanks for merging, @loganharbour

**Pull request merged**
